### PR TITLE
[AIP-51] Executors vending CLI commands

### PR DESCRIPTION
**Pull request by o-nikolas**
This PR addresses executor coupling **5a** and **5b** in Airflow CLI. See [this issue](https://github.com/apache/airflow/issues/27932) (#27932) for more context.

## Please read before reviewing: 
Change summary:

1. Updates to the executor interface (i.e. add stub methods to the base executor and the two hybrid executors which don't inherit directly from base) to vend cli commands and make small changes in the cli_parser.py module to read in commands from this new method. Add tests for the interface.

 1. Move executor related cli commands (for Celery and Kubernetes) to their respective executor modules utilizing the new interface from step 2. All original tests for these CLI commands still apply.
 
## NOTE

After this PR, CLI commands will only be displayed for the **currently configured executor**. This solves some technical issues around parsing and import cycles, as well as finally puts to rest many strange behaviours people would have with using CLI commands for executors they do not have installed which often yielded broken behaviour due to missing dependencies.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/cli/cli_parser.py
Comment (o-nikolas) [inline airflow/cli/cli_parser.py]: The implication of moving the executor cli commands to the executor modules is that if the executor module fails to load, then its cli commands won't be included in Help output. I think this is actually a fair behaviour for third party executors and even the core executors that are built into airflow. However, it is not the current behavior we have today, which is that the executor cli commands will still show up in help if they are not importable, but you'll get either warnings or stacktraces when you actually try to use them (since at that point you'll hit the importing issues). We try to catch those traces and log in this [`DefaultHelpParser._check_value`](https://github.com/aws-mwaa/upstream-to-airflow/blob/689cd36ec99fc01ab74a24b08ad78c793889feae/airflow/cli/cli_config.py#L61-L97) method, but it only looks for the celery dependency (and there could be many other ways that the module fails to load).

I'm not yet personally decided on what the right approach is here. I'd appreciate input from other folks

Comment (o-nikolas) [reply airflow/cli/cli_parser.py]: One way to mitigate this, in a generalized way, would be to catch import errors for the executors as we loop through them looking for CLI commands, and then log to the user that they could not be imported, which _may_ mean some cli commands are not available for those executors and the user should first fix the issues if they wish to use those commands.

@potiuk @pierrejeambrun @eladkal any thoughts?

**Comment by potiuk** [comment]
(sorry for looong delay). Yes. All that is cool - and I think we should simply split it into three PRs to do a final review :)

**Comment by o-nikolas** [comment]
> (sorry for looong delay). Yes. All that is cool - and I think we should simply split it into three PRs to do a final review :)

Thanks for checking in on this one @potiuk. And that's a fair callout. I usually try to keep PRs small and tactical, but this one felt cohesive. BUT, I think I can at least separate the first comment from the second two and they would still make sense to merge that way.

BTW, did you have a read through the comment thread above about the CLI Help output? I'm curious to hear your perspective on that.

**Review by potiuk — APPROVED** [review]

File: airflow/cli/cli_parser.py
Comment (potiuk) [reply airflow/cli/cli_parser.py]: This is fine. We already have precedent which is similar but even worse. When we have no CeleryExecutor **configured**, running "celery" command group yields this:

```
[jarek:~/code/airflow] [airflow-3.7] fix-infinite-nofiles-from-containerd+ 2s 130 ± airflow celery            
Usage: airflow [-h] GROUP_OR_COMMAND ...

Positional Arguments:
  GROUP_OR_COMMAND

    Groups:
      celery         Celery components
      config         View configuration

.....

Optional Arguments:
  -h, --help         show this help message and exit

airflow command error: argument GROUP_OR_COMMAND: celery subcommand works only with CeleryExecutor, CeleryKubernetesExecutor and executors derived from them, your current executor: SequentialExecutor, subclassed from: BaseExecutor, see help above.
```

I found it annoying when testing some stuff but I never got to a good solution. And when I think about it - I am more annoyed by the fact that I see the command group but I cannot actually see what subcommands it has than anything else. And this is more of a maintainer problem than user problem. 

And I even thik in this case your proposal combined with moving the celery executors to providers - we could get rid of the above warning (or rather move it to each of the subcommands) and that would solve the "annoyance" I had. This way you would only see celery command when you have "celery provider" installed (once we move the executor to celery) and not when you have it configured (and I'd love it). This is IMHO much more consistent approach

I think this is also a nice way to actually finally decouple executors - this is a nice thing that the built-in executor CLIs will add new CLI commands.

We still have the documenttion generated for all the built-in executors, so even if they are missing in the main CLI that's more than enough IMHO.

**Comment by potiuk** [comment]
BTW. It's ok to merge it as a single squashed PR I think

File: airflow/cli/cli_parser.py
Comment (o-nikolas) [reply airflow/cli/cli_parser.py]: I really appreciate the time to read the thread and send a thoughtful reply @potiuk :pray: 

Unless anyone else disagrees I'll go ahead with this approach. 
There will also need to be a followup PR which removes some of the error handling we attempt to do (basically the thing that produces that message you ran into). Core airflow should not be attempting to do this, I think we agree on that, but correct me if not.

**Comment by potiuk** [comment]
Go ahead I think @o-nikolas :)

File: airflow/cli/cli_parser.py
Comment (potiuk) [reply airflow/cli/cli_parser.py]: > Core airflow should not be attempting to do this, I think we agree on that, but correct me if not.

Yes. we agree.

**Comment by o-nikolas** [comment]
> BTW. It's ok to merge it as a single squashed PR I think

Hey @potiuk 

I had already started pulling out the commit which refactors `cli_parser.py` into it's own PR which can be reviewed here #29962


We can review and approve that PR and then once it's merged, I'll rebase this PR (which will just represent the Executor related changes). I think that will be cleaner from a git log perspective. Since the two sets of changes are related, but mostly mutually exclusive. 

Thoughts?

**Comment by potiuk** [comment]
> I had already started pulling out the commit which refactors cli_parser.py into it's own PR which can be reviewed here https://github.com/apache/airflow/pull/29962

Cool. Let's :) (will take a look but not sure if that shortly).

**Comment by o-nikolas** [comment]
Looks like the changes in #29569 have regressed this code as well (we're now too eager to fail on sqlite compat, when in the situation of CLI, we don't really care). I'll need to figure something out to fix this.

**Comment by potiuk** [comment]
Fix to failing tests (I think) in https://github.com/apache/airflow/pull/29608

**Comment by potiuk** [comment]
You can rebase now.

**Comment by o-nikolas** [comment]
Thanks @potiuk, I'll rebase now.

I have also noticed that since we now load the executors to retrieve CLI commands which they want to vend, it does slow down the time it takes to run a CLI command (importing the executors is expensive, they have quite a large dependency trees). I'll try collect some data today and post more information with this freshly rebased code.

**Comment by potiuk** [comment]
> Thanks @potiuk, I'll rebase now.
> 
> I have also noticed that since we now load the executors to retrieve CLI commands which they want to vend, it does slow down the time it takes to run a CLI command (importing the executors is expensive, they have quite a large dependency trees). I'll try collect some data today and post more information with this freshly rebased code.

Yeah. I think the error was really indication of this problem (so in a way it's good it failed). I think we do not really import all of that before (we are doing a lot of lazy loading to speed up CLI commands). 

Maybe we should actually add some tests for CLIs being **fast** (@ashb @uranusjr WDYT? - you've been doing a lot of lazy loading recently and in the past )

**Comment by o-nikolas** [comment]
This is still blocked by #30727 waiting for a merge of that one

**Comment by potiuk** [comment]
> This is still blocked by #30727 waiting for a merge of that one

Maybe good time to merge it.

**Comment by o-nikolas** [comment]
> > This is still blocked by #30727 waiting for a merge of that one
> 
> Maybe good time to merge it.

File: airflow/executors/celery_kubernetes_executor.py
Comment (jedcunningham) [inline airflow/executors/celery_kubernetes_executor.py]: Hmm, not sure this is right? Don't we want to expose the celery and k8s cli still?

File: airflow/executors/executor_loader.py
Comment (jedcunningham) [inline airflow/executors/executor_loader.py]:
```suggestion
        cls, executor_name: str, validate: bool = True
```

File: airflow/executors/executor_loader.py
Comment (jedcunningham) [inline airflow/executors/executor_loader.py]:
```suggestion
    def import_default_executor_cls(cls, validate: bool = True) -> tuple[type[BaseExecutor], ConnectorSource]:
```

File: airflow/executors/executor_loader.py
Comment (jedcunningham) [inline airflow/executors/executor_loader.py]:
```suggestion
    def import_all_executor_classes(cls, *, validate: bool = True) -> list[type[BaseExecutor]]:
```

kwarg only probably makes sense here?

File: airflow/executors/local_kubernetes_executor.py
Comment (jedcunningham) [inline airflow/executors/local_kubernetes_executor.py]: Same here, this should include the k8s cli?

File: airflow/executors/celery_kubernetes_executor.py
Comment (o-nikolas) [reply airflow/executors/celery_kubernetes_executor.py]: This has moved to the celery provider packages since this revision was out. I suppose the right thing here is to expose the celery CLI commands.

File: airflow/executors/executor_loader.py
Comment (o-nikolas) [reply airflow/executors/executor_loader.py]: After some further discussion with @potiuk we're only going to load and inject CLI commands for the currently configured executor, so I actually no longer need this method. I'll just scrap it altogether for now.

Comment (o-nikolas) [reply airflow/executors/executor_loader.py:196]: Actually, I'll include both celery and kube. I think it's safe to assume that dependencies for both should be available.

File: airflow/executors/celery_kubernetes_executor.py
Comment (o-nikolas) [reply airflow/executors/celery_kubernetes_executor.py]: Actually, I'll include both celery and kube. I think it's safe to assume that dependencies for both should be available even if this code lives within one provider, but it does create a cross provider dependency (but I think by virtue of using this executor the user has accepted that).

File: airflow/cli/cli_config.py
Comment (o-nikolas) [inline airflow/cli/cli_config.py:64]: All of this was an attempt to handle the situation of using CLI commands with missing deps. Now only the configured executor commands are available and we assume that the dependencies for that are installed (otherwise the user has much bigger issues to worry about).

File: airflow/cli/cli_parser.py
Comment (o-nikolas) [inline airflow/cli/cli_parser.py:54]: Under no circumstances should the executor commands fail the parsing process. So except any Exception here.

File: airflow/stats.py
Comment (o-nikolas) [inline airflow/stats.py:25]: These are incredibly expensive imports, especially otel, which were slowing down the CLI parser. Import them below only where they are needed.

File: tests/cli/conftest.py
Comment (o-nikolas) [inline tests/cli/conftest.py:37]: Add some more fake executors to test with below.

File: tests/cli/test_cli_parser.py
Comment (o-nikolas) [inline tests/cli/test_cli_parser.py:209]: Simplify this test, no need to test every sub command, also add kubernetes

File: airflow/cli/cli_parser.py
Comment (potiuk) [inline airflow/cli/cli_parser.py]: I think it would be good to at least print the mesage of the exception that caused it. Otherwise the user will be completely unaware what might be the reason and when the user raises issue "my airflow does not work, what should I do?"  - not even the user but also whoever looks at it in GitHub issues will have no clue either, because the actual exception will be swallowed and we will have no clue how to help the user.

Ideally we should also separately handle ImportError specifically, separately. Having an ImportError is a clear indication that the user has no right provider installed or no provider at alll. This will be probably 9X% of import errors here, so we can give the user a much more specific error:

1) first of all print the exact exception message that might give the exact import error.
2) We can add a generic advice: "If your executor is based on Celery - you need to install 3.3.0+ version of provider, if it is Kubernetes - yoy need to install 7.4.0+ of cncf.kubernetes. While we cannot be sure if this is the reason, this will be it in vast majority of cases and might significantly incrase the number of cases where the users will be able to install provider and solve the problem on their own rather than opening an issue to Airflow.

File: airflow/providers/celery/executors/celery_executor.py
Comment (potiuk) [inline airflow/providers/celery/executors/celery_executor.py]: Should it be list[GroupCommand]?

File: airflow/executors/base_executor.py
Comment (potiuk) [inline airflow/executors/base_executor.py]: I guess list[GroupCommand] or list[GroupCommand]. This is part of the public API, I think not specifying this leaves a lot of room for interpretation, what it could be. I think it only makes sense that executor adds the whole "Group" rather than single command (very similar as in case of config contributed for provider where we decided only whole section can be contributed by provider).

File: tests/providers/celery/executors/test_celery_kubernetes_executor.py
Comment (potiuk) [inline tests/providers/celery/executors/test_celery_kubernetes_executor.py]: Should it be here and in others with "assert" rather than "assert not" - `test_cli_commands_vended` instead ?

**Comment by potiuk** [comment]
Added my comments - looks good, but I think we should slightly improve diagnostics of problems and typing hints.

**Comment by potiuk** [comment]
BTW. I really like how it makes the core even more "lean" without provider's discovery. That's really good observation that we ONLY need chosen executor's command set and "other's" are not needed.

BTW2. You will also have to handle conditional imports of the `cli_config` which apparently was not present in Airflow 2.4. This is a bit unneeded because the _executor* classes are going to be only used in Airflow 2.7, so maybe the best way to handle that would be to exclude them from the import check we do when verifying provider's 2.4 compatibility. 

The easiest way to handle it @o-nikolas - is to try/except Import error the imports and check if airfow version is < 2.7.0 and raise AirflowOptionalProviderFeatureException.  The AirflowOptionalProviderFeatureException is ignored when we verify if providers are OK for old versions of airflow.

File: airflow/cli/cli_parser.py
Comment (o-nikolas) [reply airflow/cli/cli_parser.py]: > I think it would be good to at least print the mesage of the exception that caused it.

Yeah, that's fair, and I did try it with this configuration (as well as printing the full stack trace) but it really makes the output of the CLI commands ugly, and if the user doesn't even care about executor commands they're going to see the exception and a large warning every time which is a bad user experience as well.

But I can try add it back in again and see how it looks.

> We can add a generic advice: "If your executor is based on Celery - you need to install 3.3.0+ version of provider, if it is Kubernetes - yoy need to install 7.4.0+ of cncf.kubernetes. While we cannot be sure if this is the reason, this will be it in vast majority of cases and might significantly incrase the number of cases where the users will be able to install provider and solve the problem on their own rather than opening an issue to Airflow.

It feels a bit wrong to me to add this much celery/kubernetes specific hardcoding into this generic executor code. I'll try update the general exception message I print here to mention the celery/kube versions, but I think we should remove those soon in the future.

File: tests/providers/celery/executors/test_celery_kubernetes_executor.py
Comment (o-nikolas) [reply tests/providers/celery/executors/test_celery_kubernetes_executor.py]: Yupp, the combination executors used to not vend commands, but I changed it to vend after a comment from Jed. I'll update the name here :+1:

**Comment by o-nikolas** [comment]
> BTW. I really like how it makes the core even more "lean" without provider's discovery. That's really good observation that we ONLY need chosen executor's command set and "other's" are not needed.

:smiley: 


> BTW2. You will also have to handle conditional imports of the `cli_config` which apparently was not present in Airflow 2.4. This is a bit unneeded because the _executor* classes are going to be only used in Airflow 2.7, so maybe the best way to handle that would be to exclude them from the import check we do when verifying provider's 2.4 compatibility.
> 
> The easiest way to handle it @o-nikolas - is to try/except Import error the imports and check if airfow version is < 2.7.0 and raise `AirflowOptionalProviderFeatureException`. The `AirflowOptionalProviderFeatureException` is ignored when we verify if providers are OK for old versions of airflow.

Okay, I'll try whip this up today. What message should we add in the `AirflowOptionalProviderFeatureException`? Or is this really only for our import checks and is unlikely to be hit by a user?

File: airflow/cli/cli_parser.py
Comment (o-nikolas) [reply airflow/cli/cli_parser.py]: Ah, actually, after reviewing this I'm realizing I actually already solved this by using `log.exception`. That includes the exception message along with the log message. So I think we're all good on that front.

But I'll add a bit more info to the following message I display to have some info about celery/kubernetes
@potiuk

Comment (o-nikolas) [reply airflow/cli/cli_parser.py]: Here is a screenshot of how it will look. Some things to note:

1. I start by uninstalling a dependency of Celery
1. Notice how if the executor is something like LocalExecutor, we get no warnings and everything works just fine :smiley: 
1. If we switch to CeleryExecutor, we get:
   1. A message saying we failed to load CLI commands from executor <executor name>
   1. The traceback and Exception message
   1. And the helpful tip (now updated with more details from your suggestion @potiuk)
1. The CLI command still works despite the issue 
  
Let me know if you think this suffices
Screenshot from 2023-07-28 13-46-38

**Review by jedcunningham — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Fantastic!

File: airflow/cli/cli_parser.py
Comment (potiuk) [reply airflow/cli/cli_parser.py]: Yeah. That loooks cool. Even if ti is a bit ugly, this is a very good message to act on by the user (and if the user will create an issue - which will inevitably happen - "my command failed, what do I do" - we can simply respond -" just follow what is written in the message" and close such issue. (And I am not complaining, this is a fact of life that no matter what you do, some users will not read those error messages - but at least they will cop&paste it so that you can read it and ask the user to read it)l

Comment (potiuk) [reply airflow/cli/cli_parser.py]: Also: note @o-nikolas how much the "specific" message is needed. 

Put yourself in the shoes of a new user, who read somewhere that in order to use Celery executor, they need to set the env variable (but did not have celery provider installed and has no idea it needs to be added - because they have not read and digested the docs, they are just experimenting).

If you **just** print the exception and no extra message, the user would be at a total loss on what to do because the import error would not be telling anythong to the user who would not know that they have to installl celery provider. They would just see import error and no further explanation. With the extra message, we make it very clear what the problem very likely is and give the user clear instruction what to do to fix the problem.

Without it, we invite the uer to simply open an issue to Airflow and say "Airflow does not work - I have that exception, what do I do ?". With this extra message we - in vast majority of cases - prevent that question even being asked because we provide an answer to it before the user asks it.

**Comment by potiuk** [comment]
> Okay, I'll try whip this up today. What message should we add in the AirflowOptionalProviderFeatureException? Or is this really only for our import checks and is unlikely to be hit by a user?

What you proposed was good :)

**Pull request merged**
